src/exodus/sigmawallet.cpp

```diff
@@ -0,0 +1,597 @@
+            }
+
+        }, false, false);
+
+    } catch (std::runtime_error const &e) {
+        LogPrintf("%s : fail to reset all mints chain state, %s\n", __func__, e.what());
+        throw;
+    }
+}
+
+bool SigmaWallet::SetMintSeedSeen(
+    MintPoolEntry const &mintPoolEntry,
+    uint32_t propertyId,
+    uint8_t denomination,
+    exodus::SigmaMintChainState const &chainState,
+    uint256 const &spendTx)
+{
+    // Regenerate the mint
+    auto const &pubcoin = mintPoolEntry.key;
+    auto const &seedId = mintPoolEntry.seedId;
+    auto seedIndex = GetSeedIndex(seedId);
+
+    SigmaMintId id(propertyId, denomination, pubcoin);
+
+    uint160 serialId;
+
+    // Can regenerate if unlocked (cheaper)
+    if (!pwalletMain->IsLocked()) {
+
+        uint512 seed;
+        GenerateSeed(seedId, seed);
+
+        SigmaPrivateKey coin;
+        if (!SeedToPrivateKey(seed, coin)) {
+            return false;
+        }
+
+        serialId = GetSerialId(coin.serial);
+    } else {
+
+        SigmaMint mint;
+        if (!CWalletDB(walletFile).ReadExodusHDMint(id, mint)) {
+            return false;
+        }
+
+        serialId = mint.serialId;
+    }
+
+    // Create mint object
+    SigmaMint mint(
+        propertyId,
+        denomination,
+        seedId,
+        serialId);
+    mint.chainState = chainState;
+    mint.spendTx = spendTx;
+
+    AddToWallet(mint);
+
+    return true;
+}
+
+SigmaMint SigmaWallet::UpdateMintChainstate(SigmaMintId const &id, SigmaMintChainState const &state)
+{
+    return UpdateMint(id, [&state](SigmaMint &m) {
+        m.chainState = state;
+    });
+}
+
+SigmaMint SigmaWallet::UpdateMintSpendTx(SigmaMintId const &id, uint256 const &tx)
+{
+    return UpdateMint(id, [&tx](SigmaMint &m) {
+        m.spendTx = tx;
+    });
+}
+
+// Mint querying
+bool SigmaWallet::HasMint(SigmaMintId const &id) const
+{
+    CWalletDB walletdb(walletFile);
+    return walletdb.HasExodusHDMint(id);
+}
+
+bool SigmaWallet::HasSerial(secp_primitives::Scalar const &scalar) const
+{
+    CWalletDB walletdb(walletFile);
+    auto serialHash = GetSerialId(scalar);
+    return walletdb.HasExodusMintID(serialHash);
+}
+
+SigmaMint SigmaWallet::GetMint(SigmaMintId const &id) const
+{
+    CWalletDB walletdb(walletFile);
+    SigmaMint m;
+    if (!walletdb.ReadExodusHDMint(id, m)) {
+        throw std::runtime_error("fail to read hdmint");
+    }
+
+    return m;
+}
+
+SigmaMint SigmaWallet::GetMint(secp_primitives::Scalar const &serial) const
+{
+    return GetMint(GetMintId(serial));
+}
+
+SigmaMintId SigmaWallet::GetMintId(secp_primitives::Scalar const &serial) const
+{
+    CWalletDB walletdb(walletFile);
+
+    SigmaMintId id;
+    auto serialHash = GetSerialId(serial);
+    if (!walletdb.ReadExodusMintID(serialHash, id)) {
+        throw std::runtime_error("fail to read id");
+    }
+
+    return id;
+}
+
+size_t SigmaWallet::ListSigmaMints(
+    std::function<void(SigmaMint &)> const &f, bool unusedOnly, bool matureOnly) const
```

Comment:
We don't need `unusedOnly` and `matureOnly` due to the caller can filter out from their callback.